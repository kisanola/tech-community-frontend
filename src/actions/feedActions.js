import server from '../Api/server';

import * as types from '../actionTypes/feed';

import { FEED_LIMIT } from '../constants';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOOGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const setFeed = (payload) => ({
  type: types.SET_FEED,
  payload,
});

/**
 * Set feed organization
 * @return {object} action
 */
export const setFeedOrganizations = (payload) => ({
  type: types.SET_FEED_ORGANIZATIONS,
  payload,
});

/**
 * Get feed
 * @return {object} response
 */
export const getFeed = ({
  limit,
  itemsLength,
  category = 'general',
  search = '',
}) => async (dispatch) => {
  if (limit > itemsLength + FEED_LIMIT) return;

  dispatch(toggleLoading(true));

  try {
    const {
      data: { feed },
    } = await server.get(
      `/api/v1/feed?offset=${itemsLength}&limit=${FEED_LIMIT}&category=${category}&search=${search}`,
    );

    dispatch(setFeed(feed));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};


/**
 * Get feed
 * @return {object} response
 */
export const getFeedOrganizations = () => async (dispatch) => {
  const res = await server.get(
    '/api/v1/organizations?limit=5',
  );
  if (res) {
    const {
      data: { organizations },
    } = res;
    dispatch(setFeedOrganizations(organizations.docs || []));
  }
};
