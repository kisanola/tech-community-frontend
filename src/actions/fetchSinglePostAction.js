import server from '../Api/server';

import * as types from '../actionTypes/singlePostTypes';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOGGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const fetchSuccess = (post) => ({
  type: types.FETCH_SUCCESS,
  payload: post,
});

/**
 * Get feed
 * @return {object} response
 */
export default (slug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    const {
      data: { post },
    } = await server.get(
      `/posts/${slug}`,
    );

    dispatch(fetchSuccess(post));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
