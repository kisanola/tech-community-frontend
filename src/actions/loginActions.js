import axios from 'axios';
import {
  LOGIN_ERROR,
  LOGIN_SUCCESS,
  LOGIN_STARTED,
  LOGIN_INPUT_CHANGED,
} from '../actionTypes/loginTypes';

/**
 * Updates the form inputs
 * @param {*} attribute
 * @return {object} action
 */
export const formInputChanged = attribute => ({
  type: LOGIN_INPUT_CHANGED,
  payload: attribute,
});

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
export const loginStarted = () => ({
  type: LOGIN_STARTED,
});

/**
 * Triggers an action
 * when the process succeed
 * @param {object} user
 * @return {object} action
 */
export const loginSuccess = user => ({
  type: LOGIN_SUCCESS,
  payload: user,
});

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */

export const loginError = error => ({
  type: LOGIN_ERROR,
  payload: error,
});

/**
 * Login the user
 * @param {*} { username, email, password }
 * @return {object} response
 */
export const login = ({ email, password }) => async dispatch => {
  dispatch(loginStarted());

  try {
    const res = await axios.post(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/login`,
      {
        username: email,
        password,
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const { token, user } = res.data;

    localStorage.setItem('token', token);

    dispatch(loginSuccess(user));
    return true;
  } catch (e) {
    if (e.response) {
      if (e.response.data.errors) {
        dispatch(loginError(e.response.data.errors[0].message));
        return false;
      }
      dispatch(loginError(e.response.data.message));
      return false;
    }
    dispatch(loginError('Please check your internet connection'));
    return false;
  }
};
