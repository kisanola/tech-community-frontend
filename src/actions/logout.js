import { LOGOUT_SUCCESS } from '../actionTypes/currentUserTypes';

export const setIsLoggedOut = () => dispatch => {
  localStorage.removeItem('username');
  localStorage.removeItem('token');
  return dispatch({
    type: LOGOUT_SUCCESS,
    payload: true,
  });
};
