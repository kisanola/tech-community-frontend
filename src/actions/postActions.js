import axios from 'axios';
import server from '../Api/server';
import {
    POST_ERROR,
    POST_STARTED,
    POST_SUCCESS
} from '../actionTypes/postTypes';

const {
    REACT_APP_CLOUDINARY_URL_IMAGE,
    REACT_APP_CLOUDINARY_PRESET,
} = process.env;

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
const postStarted = () => ({ type: POST_STARTED })

/**
 * Triggers an action
 * when the process succeed
 * @param {object} post
 * @return {object} action
 */
const postSuccess = (post) => ({
    type: POST_SUCCESS,
    payload: post
})

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */
const postError = (error) => ({
    type: POST_ERROR,
    payload: error
})

const uploadImage = async file => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', REACT_APP_CLOUDINARY_PRESET);

    try {
        const res = await axios({
            url: REACT_APP_CLOUDINARY_URL_IMAGE,
            method: 'POST',
            header: {
                'Content-Type': 'application/x-ww-form-urlencoded',
            },
            data: formData,
        })
        return res.data.secure_url

    } catch (error) {
        throw error;
    }
};


/**
 * post
 * @param {*} { username, email, password }
 * @return {object} response
 */
const post = (data) => async dispatch => {
    dispatch(postStarted());

    try {
        const payload = {
            description: data.value,
            title: 'this is a wrong title'
        };

        if (data.image) payload.image = await uploadImage(data.image);

        const res = await server.post(`/api/v1/posts`, payload);

        const { post } = res.data;
        dispatch(postSuccess(post));
        return true;
    } catch (e) {
        if (e.response) {
            if (e.response.data.errors) {
                dispatch(postError('The post should have at least 50 characters long'));
                return false;
            }
            dispatch(postError(e.response.data.message));
            return false;
        }
        dispatch(postError('Please check your internet connection'));
        return false;
    }
};

export default post;

