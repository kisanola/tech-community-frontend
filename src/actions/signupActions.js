import axios from 'axios';
import {
  SIGNUP_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_STARTED,
  SIGNUP_INPUT_CHANGED,
} from '../actionTypes/signupTypes';

/**
 * Updates the form inputs
 * @param {*} attribute
 * @return {object} action
 */
export const formInputChanged = attribute => ({
  type: SIGNUP_INPUT_CHANGED,
  payload: attribute,
});

/**
 * Triggers an action
 * when the sign up process starts
 * @return {object} action
 */
export const signupStarted = () => ({
  type: SIGNUP_STARTED,
});

/**
 * Triggers an action
 * when the sign up process succeed
 * @param {object} user
 * @return {object} action
 */
export const signupSuccess = user => ({
  type: SIGNUP_SUCCESS,
  payload: user,
});

/**
 * Triggers an action
 * when the sign up process fails
 * @param {object} error
 * @return {object} action
 */

export const signupError = error => ({
  type: SIGNUP_ERROR,
  payload: error,
});

/**
 * Register the user
 * @param {*} { username, email, password }
 * @return {object} response
 */
export const signup = ({
  companyName,
  username,
  email,
  password,
}) => async dispatch => {
  dispatch(signupStarted());

  try {
    const res = await axios.post(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/signup`,
      {
        companyName,
        username,
        email,
        password,
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const { token, user } = res.data;

    localStorage.setItem('token', token);

    dispatch(signupSuccess(user));
    return true;
  } catch (e) {
    if (e.response) {
      if (e.response.data.errors) {
        dispatch(
          signupError(`The password should have uppercase letters and numbers`),
        );
        return false;
      }
      dispatch(signupError(e.response.data.message));
      return false;
    }
    dispatch(signupError('Please check your internet connection'));
    return false;
  }
};
