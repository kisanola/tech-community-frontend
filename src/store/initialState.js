module.exports = {
  currentUser: {
    isAuth: localStorage.getItem('token') !== null,
    user: { username: null },
  },
  feed: {
    items: [],
  },
  search: {
    searchKeywords: '',
    searchItems: [],
    searchState: {
      query: '',
    },
  },
  posts: {},
  sharePost: {
    show: false,
    content: {
      url: '',
      title: '',
      content: '',
    },
  },
  singlePost: {
    post: {
      author: {},
      slug: '',
      userType: '',
      description: '',
      likesCount: 0,
      createdAt: '',
    },
    relatedPosts: {
      isEmpty: false,
      feed: [],
    },
  },
  socialAuth: { show: false },
};
