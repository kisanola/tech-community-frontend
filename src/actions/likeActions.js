import server from '../Api/server';

import { toggleLoading } from './feedActions';

/**
 * Like a post
 * @return {object} action
 */
export const like = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.post(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};

/**
 * Unlike a post
 * @return {object} action
 */
export const unlike = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.delete(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
