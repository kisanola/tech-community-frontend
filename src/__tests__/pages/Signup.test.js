import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import axios from 'axios';
import moxios from 'moxios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { signup as signupAction } from '../../actions/signupActions';
import signupReducer from '../../reducers/signup';
import {
  SIGNUP_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_STARTED,
} from '../../actionTypes/signupTypes';
import {
  Signup,
  mapStateToProps,
  mapDispatchToProps,
} from '../../pages/Signup/Signup';
import initialState from '../../store/initialState';

let wrapper;

const signup = () => {
  return Promise.resolve(true);
};

const props = {
  signup,
  onFormInputChanged: data => null,
  history: {
    push(val) {
      return jest.fn(val);
    },
  },
};

const user = {
  organizationName: 'name',
  username: 'username',
  email: 'email@email.com',
  password: 'password',
};

describe('Signup.jsx', () => {
  beforeEach(() => {
    wrapper = mount(
      <Router>
        <Signup {...props} />
      </Router>,
    );

    moxios.install(axios);
  });

  afterEach(() => {
    moxios.uninstall(axios);
  });

  test('should render Signup.jx', () => {
    wrapper = shallow(<Signup />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should signup the user', () => {
    moxios.stubRequest(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/signup`,
      {
        status: 201,
      },
    );

    wrapper.find('.register').simulate('click');

    wrapper
      .find('input[name="companyName"]')
      .simulate('change', { target: { value: 'name' } });
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: 'username' } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: 'username@email.com' } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: 'wrongformat' } });

    wrapper.find('button').simulate('click');

    shallow(<Signup {...props} />)
      .instance()
      .onSubmit();
    expect(wrapper.props().children.props.history.push).toEqual(
      expect.any(Function),
    );
  });

  describe('reducers', () => {
    test('should return `mapStateToProps`', () => {
      const expectedState = {
        loading: false,
        error: null,
      };
      const state = mapStateToProps(initialState);
      expect(state).toEqual(expectedState);
    });
    test('should return `mapDispatchToProps`', () => {
      const dispatch = action => action;
      const mappedObject = mapDispatchToProps(dispatch);
      expect(mappedObject).toHaveProperty('signup');
      expect(mappedObject.signup('user')).toEqual(expect.any(Function));
    });
  });
});

const mockStore = configureMockStore([thunk]);

describe('Signup actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      events: {},
    });

    moxios.install(axios);
  });

  afterEach(() => {
    moxios.uninstall(axios);
  });

  it('should call the signup action with success', () => {
    moxios.stubRequest(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/signup`,
      {
        status: 201,
        response: {
          data: {
            user,
            token: 'token',
          },
        },
      },
    );
    store.dispatch(signupAction(user)).then(res => {
      expect(res).toBe(true);
    });
  });

  it('should call the signup action with error', () => {
    moxios.stubRequest(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/signup`,
      {
        status: 400,
        response: {
          data: {
            error: 'The password should have uppercase letters and numbers',
          },
        },
      },
    );
    store.dispatch(signupAction(user)).then(res => {
      expect(res.data.error).toBe(
        'The password should have uppercase letters and numbers',
      );
    });
  });
});

describe('Signup reducer', () => {
  it('should test the reducer with type SIGNUP_ERROR', () => {
    const state = signupReducer({}, { type: SIGNUP_ERROR, payload: {} });
    expect(state.loading).toBe(false);
  });

  it('should test the reducer with type SIGNUP_SUCCESS', () => {
    const state = signupReducer({}, { type: SIGNUP_SUCCESS, payload: {} });
    expect(state.loggedIn).toBe(true);
  });

  it('should test the reducer with type SIGNUP_STARTED', () => {
    const state = signupReducer({}, { type: SIGNUP_STARTED, payload: {} });
    expect(state.loading).toBe(true);
  });

  it('should test the reducer with type default', () => {
    const state = signupReducer({}, { type: 'default', payload: {} });
    expect(state).toEqual(expect.any(Object));
  });
});
