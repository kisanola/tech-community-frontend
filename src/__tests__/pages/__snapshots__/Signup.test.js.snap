// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup.jsx should render Signup.jx 1`] = `ShallowWrapper {}`;
